tests/drivers/clock_control/clock_control_api/src/test_clock_control.c

```diff
@@ -0,0 +1,310 @@
+	zassert_equal(0, err, "%s: Unexpected err (%d)", dev_name, err);
+
+	status = clock_control_get_status(dev, NULL);
+	zassert_equal(CLOCK_CONTROL_STATUS_OFF, status,
+			"%s: Unexpected status (%d)", dev_name, status);
+}
+
+static void test_on_off_status(void)
+{
+	test_all_instances(test_on_off_status_instance, NULL);
+}
+
+/*
+ * Test validates that if number of enabling requests matches disabling requests
+ * then clock is disabled.
+ */
+static void test_multiple_users_instance(const char *dev_name, u32_t startup_us)
+{
+	struct device *dev = device_get_binding(dev_name);
+	enum clock_control_status status;
+	int users = 5;
+	int err;
+
+	status = clock_control_get_status(dev, NULL);
+	zassert_equal(CLOCK_CONTROL_STATUS_OFF, status,
+			"%s: Unexpected status (%d)", dev_name, status);
+
+	for (int i = 0; i < users; i++) {
+		err = clock_control_on(dev, NULL);
+		zassert_equal(0, err, "%s: Unexpected err (%d)", dev_name, err);
+	}
+
+	status = clock_control_get_status(dev, NULL);
+	zassert_true(status != CLOCK_CONTROL_STATUS_OFF,
+			"%s: Unexpected status (%d)", dev_name, status);
+
+	for (int i = 0; i < users; i++) {
+		err = clock_control_off(dev, 0);
+		zassert_equal(0, err, "%s: Unexpected err (%d)", dev_name, err);
+	}
+
+	status = clock_control_get_status(dev, NULL);
+	zassert_true(status == CLOCK_CONTROL_STATUS_OFF,
+			"%s: Unexpected status (%d)", dev_name, status);
+}
+
+static void test_multiple_users(void)
+{
+	test_all_instances(test_multiple_users_instance, NULL);
+}
+
+static bool async_capable(const char *dev_name)
+{
+	struct device *dev = device_get_binding(dev_name);
+
+	if (clock_control_async_on(dev, 0, NULL) != 0) {
+		return false;
+	}
+
+	clock_control_off(dev, 0);
+
+	return true;
+}
+
+/*
+ * Test checks that callbacks are called after clock is started.
+ */
+static void clock_on_callback(struct device *dev, void *user_data)
+{
+	bool *executed = (bool *)user_data;
+
+	*executed = true;
+}
+
+static void test_async_on_instance(const char *dev_name, u32_t startup_us)
+{
+	struct device *dev = device_get_binding(dev_name);
+	enum clock_control_status status;
+	int err;
+	bool executed1 = false;
+	bool executed2 = false;
+	struct clock_control_async_data data1 = {
+		.cb = clock_on_callback,
+		.user_data = &executed1
+	};
+	struct clock_control_async_data data2 = {
+		.cb = clock_on_callback,
+		.user_data = &executed2
+	};
+
+	status = clock_control_get_status(dev, NULL);
+	zassert_equal(CLOCK_CONTROL_STATUS_OFF, status,
+			"%s: Unexpected status (%d)", dev_name, status);
+
+	err = clock_control_async_on(dev, 0, &data1);
+	zassert_equal(0, err, "%s: Unexpected err (%d)", dev_name, err);
+
+	err = clock_control_async_on(dev, 0, &data2);
+	zassert_equal(0, err, "%s: Unexpected err (%d)", dev_name, err);
+
+	/* wait for clock started. */
+	k_busy_wait(startup_us*1.5);
+
+	zassert_true(executed1, "%s: Expected flag to be true", dev_name);
+	zassert_true(executed2, "%s: Expected flag to be true", dev_name);
+}
```

Comment:
IMHO we should also check if the clocks are really on.

Reply:
It's hard to check that in generic way. Callback is the indicator